File: VMLib/Graph.cs
using System.Collections.Generic;

namespace VMLib {

    // directed graph class
    public class Graph<T> where T : class {
        public List<T> Nodes = new List<T>();
        public List<Edge> Edges = new List<Edge>();
        
        public class Edge {
            public T Start;
            public T End;
        }

        // only add node if it is not already in the graph
        public void AddNode(T node) {
            if(!Nodes.Contains(node)) {
                Nodes.Add(node);
            }
        }

        // add edge to the graph and add the nodes if they are not already in the graph
        public void AddEdge(T start, T end) {
            Edges.Add(new Edge() { Start = start, End = end });
            if(!Nodes.Contains(start)) {
                Nodes.Add(start);
            }
            if(!Nodes.Contains(end)) {
                Nodes.Add(end);
            }
        }

        // create a list of all nodes that do not have any edges ending at that node
        public List<T> NodesNoInput() {
            var ret = new List<T>();
            foreach(var node in Nodes) {
                var incoming = false;
                foreach(var edge in Edges) {
                    if(edge.End == node) {
                        incoming = true;
                    }
                }
                if(!incoming) {
                    ret.Add(node);
                }
            }
            return ret;
        }

        // create a list containing the nodes sorted in topological order
        // without modifying the graph
        public List<T> TopologicalSort() {
            // currently uses kahn's algorithm (https://en.wikipedia.org/wiki/Topological_sorting#Kahn%27s_algorithm)
            // which modifies the graph, so create list of current nodes and edges
            // so they can be restored at the end of the function
            var n = new List<T>(Nodes);
            var e = new List<Edge>(Edges);

            // sorted nodes
            var ret = new List<T>();


            List<T> s = NodesNoInput();
            while(s.Count > 0) {
                var node = s[0];
                ret.Add(node);
                RemoveNode(node);
                s = NodesNoInput();
            }

            // nodes still remaining in graph, so not all were
            // sorted - restore nodes and edges and return empty list
            if(Nodes.Count > 0) {
                Nodes = n;
                Edges = e;
                VMCore.Log.Warn("Cyclic Microcode graph: did not execute");
                return new List<T>();
            }

            Nodes = n;
            Edges = e;
            return ret;
        }

        // remove node from graph and remove all edges that link to that node
        public void RemoveNode(T node) {
            for(int i = Edges.Count - 1; i >= 0; i--) {
                var edge = Edges[i];
                if(edge.Start == node || edge.End == node) {
                    Edges.RemoveAt(i);
                }
            }
            Nodes.Remove(node);
        }
    }
}

File: microasm/src/shared/graph.c
#include "shared/graph.h"
#include "shared/platform.h"
#include <stdlib.h>

void InitGraph(Graph* graph) {
    ARRAY_ALLOC(Node, *graph, node);
    ARRAY_ALLOC(Edge, *graph, edge);
}

static bool isInArray(Node* arr, unsigned int count, Node value) {
    for(unsigned int i = 0; i < count; i++) {
        if(arr[i].value == value.value) {
            return true;
        }
    }
    return false;
}

Node* AddNode(Graph* graph, unsigned int node) {
    Node newNode = {.value = node, .removed = false};
    if(!isInArray(graph->nodes, graph->nodeCount, newNode)) {
        PUSH_ARRAY(void*, *graph, node, newNode);
        return &graph->nodes[graph->nodeCount - 1];
    }
    for(unsigned int i = 0; i < graph->nodeCount; i++) {
        if(graph->nodes[i].value == node) {
            return &graph->nodes[i];
        }
    }
    return NULL;
}

void AddEdge(Graph* graph, unsigned int start, unsigned int end) {
    PUSH_ARRAY(Edge, *graph, edge, ((Edge){
        .start = AddNode(graph, start), 
        .end = AddNode(graph, end)
    }));
}

NodeArray NodesNoInput(Graph* graph) {
    NodeArray ret;
    ARRAY_ALLOC(Node*, ret, node);
    
    for(unsigned int i = 0; i < graph->nodeCount; i++) {
        bool incoming = false;
        Node* node = &graph->nodes[i];
        if(node->removed) {
            continue;
        }
        for(unsigned int j = 0; j < graph->edgeCount; j++) {
            Edge* e = &graph->edges[j];
            if(!e->start->removed && !e->end->removed && e->end->value == node->value) {
                incoming = true;
            }
        }
        if(!incoming) {
            PUSH_ARRAY(Node*, ret, node, node);
        }
    }
    return ret;
}

NodeArray TopologicalSort(Graph* graph) {
    NodeArray ret;
    ARRAY_ALLOC(Node*, ret, node);

    NodeArray s = NodesNoInput(graph);
    while(s.nodeCount > 0) {
        Node* node = s.nodes[0];
        PUSH_ARRAY(Node, ret, node, node);
        node->removed = true;
        s = NodesNoInput(graph);
    }

    bool retNull = false;
    for(unsigned int i = 0; i < graph->nodeCount; i++) {
        if(!graph->nodes[i].removed) {
            retNull = true;
        } else {
            graph->nodes[i].removed = false;
        }
    }

    if(retNull) {
        cErrPrintf(TextRed, "Graph Cycle\n");
        exit(1);
    }

    return ret;
}